### PR TITLE
For issue #960

### PR DESCRIPTION
**Pull request by intech**
Fix https://github.com/ruflin/Elastica/issues/960

### CODE REVIEW

**Comment by ruflin** [comment]
@intech Thanks a lot for the contribution. Could you update the CHANGELOG.md file and add some basic tests?

**Comment by intech** [comment]
@ruflin okey! No problems.

**Comment by intech** [comment]
@ruflin Done!

**Comment by ruflin** [comment]
The implementation looks good to me. Just to be one the same page: As far as I understand you didn't implement the full version we discussed here: https://github.com/ruflin/Elastica/issues/960 For example type and index or not set on the document, right? The current implementation is fine for me as it creates the base for future improvements. Just to clarify that this will not directly solve what @yehosef initially planned.

**Comment by intech** [comment]
@ruflin all work perfect!
```php
$doc->set('name', 'value');
$index->updateDocument($doc);
```

**Comment by webdevsHub** [comment]
`_id`, `_type`, `_index` and `_version` works with `$doc->setParams($result->getHit())`.

There might be strange behavior if `highlight` or `_explanation` is set and moved to document. You can constrcut a result with bascially everything, `new \Elastica\Result(['some' => 1, 'wired' => 2, 'keys' => 3]);`.

File: lib/Elastica/Result.php [hunk 1]
```diff
@@ -186,6 +186,21 @@ public function getExplanation()
...
+     * Returns Document
+     * 
+     * @return \Elastica\Document
+     */
+    public function getDocument()
+    {
+        $doc = new \Elastica\Document();
+        $doc->setData($this->getSource());
+        $hit = $this->getHit();
+        unset($hit['_source']);
```
Comment (webdevsHub) [inline lib/Elastica/Result.php]: `_source` might not be set in result, so this line might throw fatal error

**Comment by webdevsHub** [comment]
An exception should be thrown if `$result->hasFields() === true` because you can not decide if `_fields` contains a full document

**Comment by intech** [comment]
@webdevsHub you have more experience than me. How to do it better?
```php
$doc->setData($this->getSource());
```
replace to `getData()` ?

```php
unset($hit['_source']);
```
replace to
```php
if($this->hasParam('_source')) unset($hit['_source']);
```

File: test/lib/Elastica/Test/ResultSetTest.php [hunk 2]
```diff
@@ -58,6 +58,35 @@ public function testArrayAccess()
...
+        $this->assertInstanceOf('Elastica\ResultSet', $resultSet);
+        
+        $documents = $resultSet->getDocuments();
+
+        $this->assertInternalType('array', $documents);
+        $this->assertEquals(3, count($documents));        
+        $this->assertInstanceOf('Elastica\Document', $documents[0]);
+        $this->assertInstanceOf('Elastica\Document', $documents[1]);
+        $this->assertInstanceOf('Elastica\Document', $documents[2]);
+        $this->assertFalse(isset($documents[3]));
```
Comment (ewgRa) [inline test/lib/Elastica/Test/ResultSetTest.php]: Make no sense, since there is assertEquals - 3 an than all three you check in next lines.

Better here to test document content, like name => 'elastica search' is really in document.

Comment (intech) [inline test/lib/Elastica/Test/ResultSetTest.php]: :+1:

File: lib/Elastica/Result.php (hunk 1 above)
Comment (ewgRa) [inline lib/Elastica/Result.php]: unset not raise notice if there is key not present in array:
php -r "\$a = array(); unset(\$a['a']);" - work without notice on 5.5.9 and I think on another php versions same.

File: test/lib/Elastica/Test/ResultSetTest.php (hunk 2 above)
Comment (intech) [inline test/lib/Elastica/Test/ResultSetTest.php]: Done!

File: lib/Elastica/Result.php (hunk 1 above)
Comment (intech) [inline lib/Elastica/Result.php]: http://php.net/manual/en/function.unset.php
> Note:
When using unset() on inaccessible object properties, the __unset() overloading method will be called, if declared.

**Comment by ruflin** [comment]
Looking at the implementation I have the gut feeling that there are quite a few edge cases which will break this implementation. But it is probably quite hard to cover all of them. To move forward I see two options:

* Merge it as it is and wait for users to come up with the edge cases (by creating and issue)
* Spend more time on the PR and improve the test set, to cover more edge cases.

I'm tempted to go with approach one but would like to get the feedback from @ewgRa and @webdevsHub here.

**Comment by intech** [comment]
@ruflin I propose to add to the whitelist \Elastica\Document with the list of params that can be there and solve this problem. And then act on the first point.

**Comment by intech** [comment]
https://github.com/elastic/elasticsearch-php/blob/master/src/Elasticsearch/Endpoints/Source/Get.php#L55

**Comment by ewgRa** [comment]
For me this PR is ok, it will work. Maybe I will implement it in different way, like DocumentBuilder::buildFromResult(Result), DocumentCollection::buildFromResult(ResultSet), but it is subjective. I just don't think add to Result and ResultSet more knowledge about Document.

I just wondering that there is #1017 that when will be implemented will provide same functionality.

**Comment by webdevsHub** [comment]
I like the whitelist approach, but whitelists always cause problems because it needs constant maintenance. Another edge case could be on disabled source field. Would be nice to have integration tests for known edge cases (insert, retrieve and update a doc with edge case mapping/structure)

**Comment by ruflin** [comment]
Lets go forward with the current state and revisit it combined with https://github.com/ruflin/Elastica/pull/1017 I agree that some great new things are ahead of us here.

@intech Could you squash your commits?

**Comment by intech** [comment]
@ruflin what should be done? do not understand :(

**Comment by ruflin** [comment]
@intech If you are not familiar with git rebase and squashing it is probably better to merge it as it is. It is easy to get it wrong the first time :-)

**Comment by ruflin** [comment]
@intech Merged. Thanks.

**Comment by intech** [comment]
@ruflin thanks! I do not understand because of his poor English, what I need to do. I was afraid to do wrong. :(

**Comment by ruflin** [comment]
@intech no worries. Rebase / Squashing can be useful to have a clean history. Here some related search results in case you are interested: https://www.google.ch/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=squash%20commits%20rebase
